src/EnergyPlus/WaterThermalTanks.cc

```diff
@@ -8965,137 +9031,28 @@ namespace WaterThermalTanks {
-                                            WaterHeaterDesuperheater(DesuperheaterNum).IterLimitErrIndex1,
-                                            PartLoadRatio,
-                                            PartLoadRatio);
-                                    }
-                                }
-                            } else if (SolFla == -2) {
-                                PartLoadRatio = max(
-                                    0.0,
-                                    min(WaterHeaterDesuperheater(DesuperheaterNum).DXSysPLR, (SetPointTemp - TankTemp) / (NewTankTemp - TankTemp)));
-                                if (!WarmupFlag) {
-                                    ++WaterHeaterDesuperheater(DesuperheaterNum).RegulaFalsiFailedNum1;
-                                    if (WaterHeaterDesuperheater(DesuperheaterNum).RegulaFalsiFailedNum1 == 1) {
-                                        ShowWarningError(WaterHeaterDesuperheater(DesuperheaterNum).Type + " \"" +
-                                                         WaterHeaterDesuperheater(DesuperheaterNum).Name + "\"");
-                                        ShowContinueError("Desuperheater unit part-load ratio calculation failed: PLR limits of 0 to 1 exceeded. "
-                                                          "Part-load ratio used = " +
-                                                          RoundSigDigits(PartLoadRatio, 3));
-                                        ShowContinueError("Please send this information to the EnergyPlus support group.");
-                                        ShowContinueErrorTimeStamp("This error occurred in heating mode.");
-                                    } else {
-                                        ShowRecurringWarningErrorAtEnd(WaterHeaterDesuperheater(DesuperheaterNum).Type + " \"" +
-                                                                           WaterHeaterDesuperheater(DesuperheaterNum).Name +
-                                                                           "\":  Part-load ratio calculation failed in heating mode warning "
-                                                                           "continues. Part-load ratio statistics follow.",
-                                                                       WaterHeaterDesuperheater(DesuperheaterNum).RegulaFalsiFailedIndex1,
-                                                                       PartLoadRatio,
-                                                                       PartLoadRatio);
-                                    }
-                                }
-                            }
-                            NewTankTemp = WaterThermalTank(WaterThermalTankNum).TankTemp;
-                        } else {
+                        // Calculate twice for consistency of desuperheater and tank source side energy transfer
+                        for (int i = 0; i < 2; i++){
```

Comment:
I added a loop of two iterations for desuperheater to get updated tank temperature (whenever desuperheater is in heat mode) and recalculate once, the energy discrepancy is cut a lot (from 1.87GJ vs 2.12GJ to 2.10GJ vs 2.13GJ). However the git diff here is messy and make it hard to review. @mjwitte @rraustad do you have any suggestion?